pinot-core/src/main/java/org/apache/pinot/core/transport/server/routing/stats/ServerRoutingStatsEntry.java

```diff
@@ -0,0 +1,94 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.transport.server.routing.stats;
+
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.pinot.common.utils.ExponentialMovingAverage;
+
+
+/**
+ *
+ *  {@code ServerRoutingStatsEntry} contains the query routing stats for a server. All access to
+ *  ServerRoutingStatsEntry should be made through ServerRoutingStatsManager.
+ */
+public class ServerRoutingStatsEntry {
+  String _serverInstanceId;
+  private final ReentrantReadWriteLock _serverLock;
+
+  // Fields related to number of in-flight requests.
+  private volatile int _numInFlightRequests;
+  private final ExponentialMovingAverage _inFlighRequestsEMA;
+
+  // Fields related to latency
+  private final ExponentialMovingAverage _latencyMsEMA;
+
+  // Hybrid score exponent.
+  private final int _hybridScoreExponent;
+
+  public ServerRoutingStatsEntry(String serverInstanceId, double alphaEMA, long autoDecayWindowMsEMA,
+      long warmupDurationMsEMA, double avgInitializationValEMA, int scoreExponent) {
+    _serverInstanceId = serverInstanceId;
+    _serverLock = new ReentrantReadWriteLock();
+
+    _inFlighRequestsEMA =
+        new ExponentialMovingAverage(alphaEMA, autoDecayWindowMsEMA, warmupDurationMsEMA, avgInitializationValEMA);
+    _latencyMsEMA =
+        new ExponentialMovingAverage(alphaEMA, autoDecayWindowMsEMA, warmupDurationMsEMA, avgInitializationValEMA);
+
+    _hybridScoreExponent = scoreExponent;
+  }
+
+  public ReentrantReadWriteLock.ReadLock getServerReadLock() {
+    return _serverLock.readLock();
+  }
+
+  public ReentrantReadWriteLock.WriteLock getServerWriteLock() {
+    return _serverLock.writeLock();
+  }
+
+  public Integer getNumInFlightRequests() {
+    return _numInFlightRequests;
+  }
+
+  public Double getInFlightRequestsEMA() {
+    return _inFlighRequestsEMA.getAverage();
+  }
+
+  public Double getLatencyEMA() {
+    return _latencyMsEMA.getAverage();
+  }
+
+  public double computeHybridScore() {
+    double estimatedQSize = _numInFlightRequests + _inFlighRequestsEMA.getAverage();
+    return Math.pow(estimatedQSize, _hybridScoreExponent) * _latencyMsEMA.getAverage();
+  }
+
+  public void updateNumInFlightRequestsForQuerySubmission() {
+    ++_numInFlightRequests;
+    _inFlighRequestsEMA.compute(_numInFlightRequests);
+  }
+
+  public void updateNumInFlightRequestsForResponseArrival() {
+    --_numInFlightRequests;
```

Comment:
Do we intentionally not updating `_inFlighRequestsEMA` here?

Reply:
Yes. It is intentional.  The reasoning is that we want a snapshot of the queue at regular intervals and just updating at querySubmit should be good enough. In the example mentioned, the ema will decrease if one of the two events happen:
1. The next query after burst is submitted. 
2. Autodecay kicks in.